src/main/java/org/fundacionjala/coding/kevinherrera/crypto/CesarCrypto.java

```diff
@@ -0,0 +1,133 @@
+     * @param key key to decode
+     * @return message as lowercase type decoded
+     */
+    public String decode(final int key, final String encodeMessage) {
+        return processEncrypt(EncryptTask.DECODE, encodeMessage, key).toLowerCase();
+    }
+
+    /**
+     * encode the alphabetic values of a message using vigenere method.
+     * @param message message to encode
+     * @param key key to encode
+     * @return message as UPPERCASE type encode
+     */
+    public String encodeVigenere(final String message, final String key) {
+        return processEncryptVigenere(EncryptTask.ENCODE, message, key);
+    }
+
+    /**
+     * decode the alphabetic values of a message encoded with vigenere key.
+     * @param message message to decode
+     * @param key key to decode
+     * @return message as lowercase type decoded
+     */
+    public String decodeVigenere(final String key, final String message) {
+        return processEncryptVigenere(EncryptTask.DECODE, message, key).toLowerCase();
+    }
+
+    /**
+     * internal process of encrypt functions.
+     * @param task encrypt task
+     * @param message message to process
+     * @param key codification key
+     * @return message encrypted or decrypted determinated by task value
+     */
+    private String processEncrypt(final EncryptTask task, final String message, int key) {
+        StringBuilder messageBuilder = new StringBuilder();
+        if (message == null) {
+            return messageBuilder.toString();
+        }
+        for (char c : message.toCharArray()) {
+            if (Character.isAlphabetic(c)) {
+                int newCharValue = 0;
+                int charHash = Character.hashCode(Character.toUpperCase(c));
+                if (task == EncryptTask.ENCODE) {
+                    int aux = charHash + Math.abs(key);
+                    int aux2 = aux - CHAR_Z_UPPER_HASHCODE;
+                    newCharValue = aux > CHAR_Z_UPPER_HASHCODE ? (CHAR_A_UPPER_HASHCODE - 1) + aux2 : charHash + key;
+                }
+                if (task == EncryptTask.DECODE) {
+                    int aux = charHash - Math.abs(key);
+                    int aux2 = (CHAR_A_UPPER_HASHCODE - 1) - aux;
+                    newCharValue = aux < CHAR_A_UPPER_HASHCODE
+                            ? CHAR_Z_UPPER_HASHCODE - (aux2) : charHash - Math.abs(key);
+                }
+                messageBuilder.append(Character.toChars(newCharValue));
+            } else {
+                messageBuilder.append(c);
+            }
+        }
+        return messageBuilder.toString();
+    }
+
+    /**
+     * internal process of encrypt using vigenere method.
+     * @param task encrypt task
+     * @param message message to process
+     * @param stringKey codification key
+     * @return message encrypted or decrypted determinated by task value
+     */
+    private String processEncryptVigenere(final EncryptTask task, final String message, final String stringKey) {

```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Refactor this method to reduce its Cognitive Complexity from 23 to the 15 allowed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3776)